feat: Topic table

HemaHG opened the pull request:
Signed-off-by: hemahg <hhg@redhat.com>

**What this PR does / why we need it**:

> Use the responsive table for the topics list view


**Which issue(s) this PR fixes** 
fixes https://issues.redhat.com/browse/MGDSTRM-9359

Link to story - https://619e1921b3e38b003afeb2be-lajvitpdte.chromatic.com/?path=/story/kafka-kafkatopics-kafkatopics--kafka-topics

riccardo-forina commented on src/Kafka/KafkaTopic/types.ts:
```diff
@@ -0,0 +1,73 @@
+
+export type KafkaTopicField = keyof KafkaTopic;
+
+export const unitsToBytes = {
+  [RetentionSizeUnits.BYTE]: 1,
+  [RetentionSizeUnits.KIBIBYTE]: 1024,
+  [RetentionSizeUnits.MEBIBYTE]: 1048576,
+  [RetentionSizeUnits.GIBIBYTE]: 1073741824,
+  [RetentionSizeUnits.TEBIBYTE]: 1.0995116e12,
+};
```
For the metrics we use an [external lib](https://github.com/redhat-developer/app-services-ui-components/blob/b97ccae1133e0ccec1c1f2ece120e8aaf8f441af/src/Kafka/Metrics/components/utils.ts#L7-L9) to convert bytes. In general, for solved problems like these, I'd prefer relying on something external (end tested) than writing some custom code ourselves.

No need to address this in this PR, but I'd make a note to streamline all these conversions. It's clear how it's a shared concern across components, so we should have some shared component/utility to deal with it.

HemaHG replied:
yeah sure, will make use of the external lib

HemaHG commented on src/Kafka/KafkaTopic/KafkaTopics.tsx:
```diff
@@ -0,0 +1,11 @@
+    topics: T[];
+    count: number;
+  }>;
+};
```
apologies @riccardo-forina, I am still working on it and didn't notice that have pushed something that's still in WIP.

HemaHG commented on src/Kafka/KafkaTopic/components/EmptyStateNoResults.tsx:
```diff
@@ -0,0 +1,24 @@
+      <EmptyStateBody>{t("common:empty_state_no_results_body")}</EmptyStateBody>
+    </EmptyState>
+  );
+};
```
moved to the shared folder, as it's used by other components like the consumer table and ACLS table

HemaHG commented:
@riccardo-forina, the existing search filter in the topic table contains search input with only chips and the value in the chips will be overwritten if we search, but the new search filter contains the chip group, can I make use of the search input component if we only have a search filter or can I makes changes in the chip filter so that it only takes a string input instead of string

HemaHG commented:
@jgiardino gave heads-up to use the new search filter

jgiardino commented on locales/en/topic.json:
```diff
@@ -1,4 +1,20 @@
+  "table": {
+    "actions": {
+      "delete": "Delete",
+      "edit": "Edit"
```
As part of refactoring the Kafka table view, we updated the labels in the action menu based on usability test findings. Can we do the same for the actions in this menu?

The updates would be:
* Edit 👉 Edit topic configuration
* Delete 👉 Delete topic

jgiardino commented on locales/en/topic.json:
```diff
@@ -1,4 +1,20 @@
   "create_topic": "Create topic",
-  "show_all_options": "Show all available options"
+  "empty_topics_body": "Create a topic by clicking the button below to get started.",
+  "empty_topics_title": "No topics yet",
```
This was one of the first UIs we implemented, and the contents have evolved slightly since this empty state was included. Is it possible to also update these contents to be consistent with others:
```
  "empty_topics_body": "To get started, create a topic.",
  "empty_topics_title": "No topics",
```

jgiardino commented:
These updates look really good! I posted a couple of text string updates inline. The only other thing I noticed is related to the filter. I'm not sure if this is addressable in this PR or how to approach this, but wanted to capture it here to figure out next steps.

In the Kafka instances table, we have many filters. One of the indicators that this UI is for filtering is the funnel/filter icon that displays in the select menu, where you select what property to filter on.

![image](https://user-images.githubusercontent.com/21063328/198692106-4984f1b7-d68c-4048-b0d7-290f99eeb0a3.png)

In the filter for topics, this select menu is not included, and therefore this additional visual indication is not included.

<img width="409" alt="image" src="https://user-images.githubusercontent.com/21063328/198692417-9ed342a7-4675-45bb-b87f-bbe51e6986d4.png">

I think the recommendation based on [what I asked in slack](https://patternfly.slack.com/archives/CGR204RM2/p1666976896041989) is that the text input should always have a search icon, whether it's preceded by a select menu or not.

This is consistent with what's shown on the filter guidelines: 
https://www.patternfly.org/v4/guidelines/filters#text-entry-filters
https://www.patternfly.org/v4/guidelines/filters#attribute-value-filter

I think this means we just need to add a 🔍 icon to the text input that displays in the filter bar, and the funnel/filter icon only displays on the select toggle. But I think this is likely a separate update from this PR. 🤷‍♀️

HemaHG commented:
sure @jgiardino, I will make [this](https://github.com/redhat-developer/app-services-ui-components/pull/546#issuecomment-1295365822) updates In another PR

jgiardino commented:
I see the updates to the text strings were made. One other thing I noticed is the alignment of the topic names

This is happening as a result of not using the inline modifier class on the link that displays in that column. There should be a prop for the Button component you can use to fix that alignment issue.

<img width="429" alt="image" src="https://user-images.githubusercontent.com/21063328/201686230-c388c771-134b-49be-a066-2834cc9e0987.png">

HemaHG commented:
@jgiardino, I have incorporated all the review comments, I have also addressed your review comment https://github.com/redhat-developer/app-services-ui-components/pull/546#issuecomment-1295365822 to include the search icon in the search input field

jgiardino reviewed (approved):
Thanks for making all the updates I suggested.

jgiardino reviewed (changes requested):
I think this is an update from this PR. When I look at [Type Error For Invalid](https://619e1921b3e38b003afeb2be-lajvitpdte.chromatic.com/?path=/story/shared-tabletoolbar-searchinput--type-error-for-invalid), I see the following:

<img width="858" alt="image" src="https://user-images.githubusercontent.com/21063328/203432115-a7971609-a36b-4bb6-87cf-91e0d3df45a6.png">

jgiardino commented:
This last comment would be a nice-to-have if it's simple, but since I didn't catch it the first time, I don't want to hold this PR up if it's a lot of effort.

One of the patterns we're trying to standardize on is to have the Delete action always be the last item in a menu. If it's possible, could you swap the order that these menu items display in?

<img width="234" alt="image" src="https://user-images.githubusercontent.com/21063328/203433227-16c83f8d-8767-4d9c-9166-57e82a3f3e4c.png">

HemaHG commented:
swapped the order of action items https://619e1921b3e38b003afeb2be-jvkoijhyux.chromatic.com/?path=/story/kafka-kafkatopics-kafkatopics--kafka-topics and removed the search filter icon, I will handle that in a separate PR

riccardo-forina reviewed (approved):
LGTM. As a next enhancement I'd like to see us start using `convert` for the bytes, as it's now a supported data type https://convert.js.org/modules#convert

jgiardino reviewed (approved):
Thanks for making the updates!

riccardo-forina commented:
:tada: This PR is included in version 2.15.0 :tada:

The release is available on [GitHub release](https://github.com/redhat-developer/app-services-ui-components/releases/tag/v2.15.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
